iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp

```diff
@@ -229,6 +236,18 @@ class VulkanSPIRVTargetBackend : public TargetBackend {
   // NOTE: we could vary this based on the options such as 'vulkan-v1.1'.
   std::string name() const override { return "vulkan*"; }
 
+  void getDependentDialects(DialectRegistry &registry) const override {
+    // clang-format off
+    registry.insert<AffineDialect,
+                    Vulkan::VulkanDialect,
+                    gpu::GPUDialect,
+                    linalg::LinalgDialect,
+                    scf::SCFDialect,
+                    spirv::SPIRVDialect,
+                    vector::VectorDialect>();
```

Comment:
This is a big layering violation. Is this the final form of this registration system? Requiring the transitive list of all dialects that may be used by any subsequent lowering is very fragile and removes a lot of the composition possibilities - someone adding a small transform through another dialect way down between vector->spirv would now have to pop back up to these files and add the deps/registration/etc. If that dialect stops being used then someone has to remember to scrub all places that may depend on it. Yuck :(

Reply:
I hope not :confused: The issue here is that because we have a pass pipeline within a pass, the pass has to say what dialects it will use up front for the entire pass pipeline. By the time runOnOperation happens, it's too late to load the dialects. There's a TODO on the original declaration of this method in the superclass referencing this, but I hope that we can revisit how to do this properly with dynamic pass pipelines. If we could walk the passes in the pipeline then we wouldn't need to declare them all here, but right now that requires a context, which we can't get in this method. I could investigate further trying to get all the passes and their dependent dialects, but I think we should just re-evaluate in the context of dynamic pass pipelines.

Reply:
Any ETA on those and an upstream LLVM bug tracking it? As above, I'm concerned that this is a big step back and something that is generally really hard to unwind the further it is allowed to persist. As with the flow->hal example breaking out-of-tree custom dialects lowering into the HAL, this change breaks out-of-tree backends lowering out of the HAL.

Reply:
https://llvm.discourse.group/t/rfc-dynamic-pass-pipeline/1637. I'm reading up on it now.

Reply:
I don't think this should break out-of-tree backends. The backend has to declare all the dependent dialects for the pass, which is gross, but it's still possible. I'm continuing the discussion about how to make this possible in the linked RFC. For now, I think this is much less bad than the ConvertFlowToHal thing, which I think I've fixed.